src/main/anovos/data_ingest/data_ingest.py

```diff
@@ -41,13 +55,129 @@ def read_dataset(spark, file_path, file_type, file_configs={}):
+            type(threshold_string) != float
+            or threshold_string < 0
+            or threshold_string > 1
+        ):
+            raise TypeError("Invalid input for threshold_string")
+        list_of_cols = list(c for c in odf.columns if c not in id_cols)
+        odf = odf.persist(pyspark.StorageLevel.MEMORY_AND_DISK)
+        odf_recast = odf.select(
+            [F.col(c).cast(T.DoubleType()).alias(c) for c in list_of_cols]
+        )
+        funs_count = [F.count]
+        exprs_count = [f(F.col(c)) for f in funs_count for c in list_of_cols]
+        list_count_recast = (
+            odf_recast.groupby().agg(*exprs_count).rdd.flatMap(lambda x: x).collect()
+        )
+
+        list_count = odf.groupby().agg(*exprs_count).rdd.flatMap(lambda x: x).collect()
+
+        funs_distinct = [F.countDistinct]
+        exprs_distinct = [f(F.col(c)) for f in funs_distinct for c in list_of_cols]
+        list_distinct_count = (
+            odf.select(*exprs_distinct).rdd.flatMap(lambda x: x).collect()
+        )
+        list_ratio = [i / j for i, j in zip(list_distinct_count, list_count)]
+        list_cat_ratio = [i / j for i, j in zip(list_count_recast, list_count)]
+
+        list_schema = []
+        if id_cols:
+            for col in id_cols:
+                list_schema.append(T.StructField(str(col), T.StringType(), True))
+        for k in range(0, len(list_count)):
+            if list_cat_ratio[k] < threshold_string:
+                list_schema.append(
+                    T.StructField(str(list_of_cols[k]), T.StringType(), True)
+                )
+            elif list_count[k] <= threshold_num:
+                if list_ratio[k] > threshold_ratio:
+                    list_schema.append(
+                        T.StructField(str(list_of_cols[k]), T.DoubleType(), True)
+                    )
+                else:
+                    list_schema.append(
+                        T.StructField(str(list_of_cols[k]), T.StringType(), True)
+                    )
+            elif (
+                list_ratio[k] < threshold_ratio
+                or list_distinct_count[k] < threshold_num
+            ):
+                list_schema.append(
+                    T.StructField(str(list_of_cols[k]), T.StringType(), True)
+                )
+
+            else:
+                list_schema.append(
+                    T.StructField(str(list_of_cols[k]), T.DoubleType(), True)
+                )
+        odf.unpersist()
+        full_schema = T.StructType(list_schema)
+        odf = (
+            spark.read.format(file_type)
+            .options(**file_configs)
+            .schema(full_schema)
+            .load(file_path)
+        )
+    if parquet_conversion:
+        file_path = parquet_output_path
+        file_type = "parquet"
+        odf.write.format(file_type).options(**file_configs).save(
+            file_path, mode="overwrite"
+        )
+        odf = spark.read.format(file_type).options(**file_configs).load(file_path)
```

Comment:
Why are you reading the dataframe for a (potentially) third time?

Reply:
The idea is to write the file in parquet_output_path in the form of parquet file, and reread it again from there.

Reply:
Does the resulting dataframe differ, though? Shouldn't it be the same data irrespective of the file type it was read from?

Reply:
As we test it out, by writing the dataframe out in parquet format, and reread the parquet dataframe again, the dataframe performs a lot better in terms of runtime compared with csv. At some cases, we see up to 50% at runtime reduction in parquet compared with csv input. The reason for this is because, Spark read and save parquet file in a different manner than csv, and parquet is optimized for big data size. As a result, that makes a huge improvement in terms of optimization.

Reply:
Does the in-memory structure of a dataframe depend on the file format it was read from? Where exactly do you see the speed-up? While reading/writing the file or while processing data in-memory?

Reply:
The speed-up we see is when the output of this function, `odf`, is being used as an input in other Anovos functions. So I would say it will be in processing data in-memory down the line.